File: 53-maximum-subarray/53-maximum-subarray.py
class Solution:    
    def maxSubArray(self, nums):
        maxSum = nums[0]
        maxVal = nums[0]
        for i in range(1, len(nums)):
            if maxVal < 0:
                maxVal = 0
            maxVal += nums[i]
            maxSum = max(maxVal, maxSum)
        return maxSum
    
    
    
    
    
    
    
    
    
    
    
    
    
#     def maxSubArray(self, nums):
#         maxVal = nums[0]
#         currVal = 0
#         for num in nums:
#             if currVal < 0:
#                 currVal = num
#             else:
#                 currVal += num
#             maxVal = max(maxVal, currVal)
#         return maxVal
    
#     # def maxSubArray(self, nums: List[int]) -> int:
#     #     current_subarray = 0
#     #     max_subarray = 0
#     #     for num in nums:
#     #         current_subarray = max(0, current_subarray + num)
#     #         max_subarray = max(current_subarray, max_subarray)
#     #     if max_subarray == 0:
#     #         return max(nums)
#     #     return max_subarray

File: 53-maximum-subarray/53-maximum-subarray.cpp
class Solution {
    public:
        int maxSubArray(vector<int>& nums){
            int maxSum = nums[0];
            int currSum = nums[0];
            for(int i = 1 ; i < nums.size() ; i++){
                if(currSum < 0)
                    currSum = 0;
                currSum += nums[i];
                maxSum = max(currSum, maxSum);
            }            
            return maxSum;
        }        
};













// class Solution {
// public:
//     int maxSubArray(vector<int>& nums) {
//         //no memo
//         int maxValue = nums[0];
//         for(int i = 1 ; i < nums.size() ; i++){
//             nums[i] = max(nums[i], nums[i]+nums[i-1]);
//             maxValue = max(maxValue, nums[i]);
//         }
//         return maxValue;
//         /*
//         DP with memo
//         vector<int> sums(nums.size());
//         sums[0] = nums[0];
//         for(int i = 1 ; i < nums.size() ; i++){
//             sums[i] = max(nums[i], sums[i-1]+nums[i]);
//         }
//         int maxValue = sums[0];
//         for(int s: sums) maxValue = max(s, maxValue);
//         return maxValue;*/ 
//     }
// };